python/message.c

```diff
@@ -0,0 +1,1487 @@
+    {Py_tp_dealloc, PyUpb_CMessage_Dealloc},
+    {Py_tp_doc, "A ProtocolMessage"},
+    {Py_tp_getattro, PyUpb_CMessage_GetAttr},
+    {Py_tp_getset, PyUpb_CMessage_Getters},
+    {Py_tp_hash, PyObject_HashNotImplemented},
+    {Py_tp_methods, PyUpb_CMessage_Methods},
+    {Py_tp_new, PyUpb_CMessage_New},
+    {Py_tp_str, PyUpb_CMessage_ToString},
+    {Py_tp_repr, PyUpb_CMessage_ToString},
+    {Py_tp_richcompare, PyUpb_CMessage_RichCompare},
+    {Py_tp_setattro, PyUpb_CMessage_SetAttr},
+    {Py_tp_init, PyUpb_CMessage_Init},
+    {0, NULL}};
+
+PyType_Spec PyUpb_CMessage_Spec = {
+    PYUPB_MODULE_NAME ".CMessage",             // tp_name
+    sizeof(PyUpb_CMessage),                    // tp_basicsize
+    0,                                         // tp_itemsize
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,  // tp_flags
+    PyUpb_CMessage_Slots,
+};
+
+// -----------------------------------------------------------------------------
+// MessageMeta
+// -----------------------------------------------------------------------------
+
+// MessageMeta is the metaclass for message objects.  The generated code uses it
+// to construct message classes, ie.
+//
+// FooMessage = _message.MessageMeta('FooMessage', (_message.Message), {...})
+//
+// (This is not quite true: at the moment the Python library subclasses
+// MessageMeta, and uses that subclass as the metaclass.  There is a TODO below
+// to simplify this, so that the illustration above is indeed accurate).
+
+typedef struct {
+  const upb_msglayout* layout;
+  PyObject* py_message_descriptor;
+} PyUpb_MessageMeta;
+
+// The PyUpb_MessageMeta struct is trailing data tacked onto the end of
+// MessageMeta instances.  This means that we get our instances of this struct
+// by adding the appropriate number of bytes.
+static PyUpb_MessageMeta* PyUpb_GetMessageMeta(PyObject* cls) {
+#ifndef NDEBUG
+  PyUpb_ModuleState* state = PyUpb_ModuleState_MaybeGet();
+  assert(!state || cls->ob_type == state->message_meta_type);
+#endif
+  return (PyUpb_MessageMeta*)((char*)cls + cpython_bits.type_basicsize);
+}
+
+static const upb_msgdef* PyUpb_MessageMeta_GetMsgdef(PyObject* cls) {
+  PyUpb_MessageMeta* self = PyUpb_GetMessageMeta(cls);
+  return PyUpb_Descriptor_GetDef(self->py_message_descriptor);
+}
+
+PyObject* PyUpb_MessageMeta_ModuleQualifiedName(const upb_msgdef* m) {
+  const upb_filedef* file = upb_msgdef_file(m);
+  const char* filename = upb_filedef_name(file);
+  const char* msgname = upb_msgdef_name(m);
+  const char* final_dot = strrchr(filename, '.');
+  size_t len = final_dot ? final_dot - filename : strlen(filename);
+  char* modname = malloc(len + 1);
+  if (!modname) return NULL;
+  for (size_t i = 0; i < len; i++) {
+    if (filename[i] == '/') {
+      modname[i] = '.';
+    } else {
+      modname[i] = filename[i];
+    }
+  }
+  modname[len] = '\0';
+  PyObject* ret = PyUnicode_FromFormat("%s_pb2.%s", modname, msgname);
+  free(modname);
```

Comment:
you know max path length from filesystem constants I think... we could avoid allocating here